python/cuda_parallel/cuda/parallel/experimental/_cccl.py

```diff
@@ -0,0 +1,209 @@
+# Copyright (c) 2024, NVIDIA CORPORATION & AFFILIATES. ALL RIGHTS RESERVED.
+#
+#
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+import numba
+import functools
+import ctypes
+from numba import types, cuda
+
+
+# MUST match `cccl_type_enum` in c/include/cccl/c/types.h
+class TypeEnum(ctypes.c_int):
+    INT8 = 0
+    INT16 = 1
+    INT32 = 2
+    INT64 = 3
+    UINT8 = 4
+    UINT16 = 5
+    UINT32 = 6
+    UINT64 = 7
+    FLOAT32 = 8
+    FLOAT64 = 9
+    STORAGE = 10
+
+
+# MUST match `cccl_op_kind_t` in c/include/cccl/c/types.h
+class OpKind(ctypes.c_int):
+    STATELESS = 0
+    STATEFUL = 1
+
+
+# MUST match `cccl_iterator_kind_t` in c/include/cccl/c/types.h
+class IteratorKind(ctypes.c_int):
+    POINTER = 0
+    ITERATOR = 1
+
+
+# MUST match `cccl_type_info` in c/include/cccl/c/types.h
+class TypeInfo(ctypes.Structure):
+    _fields_ = [
+        ("size", ctypes.c_int),
+        ("alignment", ctypes.c_int),
+        ("type", TypeEnum),
+    ]
+
+
+# MUST match `cccl_op_t` in c/include/cccl/c/types.h
+class Op(ctypes.Structure):
+    _fields_ = [
+        ("type", OpKind),
+        ("name", ctypes.c_char_p),
+        ("ltoir", ctypes.c_char_p),
+        ("ltoir_size", ctypes.c_int),
+        ("size", ctypes.c_int),
+        ("alignment", ctypes.c_int),
+        ("state", ctypes.c_void_p),
+    ]
+
+
+# MUST match `cccl_iterator_t` in c/include/cccl/c/types.h
+class Iterator(ctypes.Structure):
+    _fields_ = [
+        ("size", ctypes.c_int),
+        ("alignment", ctypes.c_int),
+        ("type", IteratorKind),
+        ("advance", Op),
+        ("dereference", Op),
+        ("value_type", TypeInfo),
+        ("state", ctypes.c_void_p),
+    ]
+
+
+# MUST match `cccl_device_reduce_build_result_t` in c/include/cccl/c/reduce.h
+class DeviceReduceBuildResult(ctypes.Structure):
+    _fields_ = [
+        ("cc", ctypes.c_int),
+        ("cubin", ctypes.c_void_p),
+        ("cubin_size", ctypes.c_size_t),
+        ("library", ctypes.c_void_p),
+        ("accumulator_size", ctypes.c_ulonglong),
+        ("single_tile_kernel", ctypes.c_void_p),
+        ("single_tile_second_kernel", ctypes.c_void_p),
+        ("reduction_kernel", ctypes.c_void_p),
+    ]
+
+
+# MUST match `cccl_value_t` in c/include/cccl/c/types.h
+class Value(ctypes.Structure):
+    _fields_ = [("type", TypeInfo), ("state", ctypes.c_void_p)]
+
+
+def _type_to_enum(numba_type):
+    mapping = {
```

Comment:
I'd move this dict out as `_TYPE_TO_ENUM_DICT` or similar, then make the function body simply:
```
    return _TYPE_TO_ENUM_DICT.get(numba_type, TypeEnum.STORAGE)
```

Reply:
Great suggestion, this also avoids reconstruction of the dictionary upon every function invocation I think. Fixed.